hack/tools/logcheck/pkg/logcheck/logcheck.go

```diff
@@ -0,0 +1,271 @@
+	var (
+		message       ast.Expr   // nil for WithValues
+		keysAndValues []ast.Expr // could be empty
+	)
+
+	switch funcName {
+	case "WithValues":
+		if len(callExpr.Args) == 0 {
+			// empty WithValues() call doesn't make any sense
+			pass.ReportRangef(selExpr.Sel, "call to %q without arguments", funcName)
+			return
+		}
+		keysAndValues = callExpr.Args
+	case "Info":
+		if len(callExpr.Args) < 1 {
+			// typecheck will complain as well, we don't need to provide helpful advice here, but we should fail nevertheless
+			pass.ReportRangef(selExpr.Sel, "call to %q is missing arguments", funcName)
+			return
+		}
+		message = callExpr.Args[0]
+		keysAndValues = callExpr.Args[1:]
+	case "Error":
+		if len(callExpr.Args) < 2 {
+			// typecheck will complain as well, we don't need to provide helpful advice here, but we should fail nevertheless
+			pass.ReportRangef(selExpr.Sel, "call to %q is missing arguments", funcName)
+			return
+		}
+		message = callExpr.Args[1]
+		keysAndValues = callExpr.Args[2:]
+	}
+
+	// first check message (if given)
+	if message != nil {
+		value, ok := isConstantStringExpr(pass, message)
+		if !ok {
+			pass.ReportRangef(message, "structured logging message should be a constant string expression")
+			return
+		}
+
+		if value == "\"\"" { // string literal or constant values contain quotes ""
+			pass.ReportRangef(message, "structured logging message should not be empty: %s", value)
+			return
+		}
+
+		// find a common pattern of mistakes: if format specifier is used in message, someone probably forgot that this is
+		// structured logging, so check for format specifier first and skip if found
+		if specifier, has := hasFormatSpecifier(value); has {
+			pass.ReportRangef(message, "structured logging message should not use format specifier %q", specifier)
+			return
+		}
+
+		if !unicode.IsUpper([]rune(value)[1]) {
+			pass.ReportRangef(message, "structured logging message should be capitalized: %s", value)
+			return
+		}
+
+		if strings.HasSuffix(value, ".\"") {
+			pass.ReportRangef(message, "structured logging message should not end with punctuation mark: %s", value)
+			return
+		}
+	}
+
+	// now check the given key-value pairs
+	checkKeysAndValues(pass, selExpr.Sel, funcName, keysAndValues)
+}
+
+func checkKeysAndValues(pass *analysis.Pass, rng analysis.Range, funcName string, keysAndValues []ast.Expr) {
+	// check even number of args
+	if len(keysAndValues)%2 != 0 {
+		pass.ReportRangef(rng, "structured logging arguments to %s must be key-value pairs, got odd number of arguments: %d", funcName, len(keysAndValues))
+		return
+	}
+
+	// check keys
+	for i, key := range keysAndValues {
+		if i%2 != 0 {
+			continue
+		}
+
+		value, ok := isConstantStringExpr(pass, key)
+		if !ok {
+			pass.ReportRangef(key, "structured logging key should be a constant string expression")
+			continue
+		}
+
+		if value == "\"\"" { // string literal or constant values contain quotes ""
```

Comment:
What about `""` (same for other occurrences)?

Reply:
The value representation can't be an empty string, I guess. If the value itself is the empty string, the representation will be `""` -> if written as a string literal `"\"\""

Reply:
Ah got it. That's a good suggestion, thanks :)